File: sinteemar/models/financa.py
from datetime import date, datetime
from sinteemar.models.arquivo import Arquivo
from sinteemar.models.usuario import Usuario

class Financa():
	'''
	Atributos:
		id (int): Chave primária
		periodo (str): Período (semanal, mensal, bimestral, trimestral, semestral ou anual)
		flag (str): Valor utilizado para ordenação por período
		documento (Arquivo): Documento
		usuario (Usuario): Usuário
		data (datetime): Data de cadastro da finança
		ativo (bool): Inativo (False) ou ativo (True)
	'''
	__slots__ = ['_ativo', '_data', '_documento', '_flag', '_id', '_periodo', '_usuario']

	def __init__(self, id: int=0, periodo: str='', flag: str=datetime.now().strftime('%Y%m'), documento: str='', usuario: Usuario=Usuario(), data: datetime=datetime.now(), ativo: bool=False):
		self._id = id
		self._periodo = periodo
		self._flag = flag
		self._documento = Arquivo(classe=self.__class__.__name__, nome=documento)
		self._usuario = usuario
		self._data = data
		self._ativo = ativo

	def __str__(self) -> str:
		return 'ID: ' + str(self._id) + '\nPERÍODO: ' + self._periodo + '\nFLAG: ' + self._flag + '\nDOCUMENTO: ' + self._documento.arquivo + '\nUSUÁRIO: ' + self._usuario.nome.title() + '\nDATA: ' + self._data.strftime('%d/%m/%Y - %H:%M:%S') + '\nATIVO: ' + str(self._ativo)

	@property
	def id(self) -> int:
		return self._id

	@id.setter
	def id(self, id: int) -> bool:
		if id > 0:
			self._id = id
			return True
		return False

	@property
	def periodo(self) -> str:
		return self._periodo

	@periodo.setter
	def periodo(self, periodo: str) -> bool:
		self._periodo = periodo.strip()
		flag = {'Jan': '01', 'Fev': '02', 'Mar': '03', 'Abr': '04', 'Mai': '05', 'Jun': '06', 'Jul': '07', 'Ago': '08', 'Set': '09', 'Out': '10', 'Nov': '11', 'Dez': '12', '1º ': '03.1', '2º ': '06.1', '3º ': '09.1', '4º ': '12.1'}
		self._flag = periodo[-4:] + flag[periodo[:3]]
		return True

	@property
	def flag(self) -> str:
		return self._flag

	@flag.setter
	def flag(self, flag: str) -> bool:
		periodos = {'01': 'Janeiro', '02': 'Fevereiro', '03': 'Março', '04': 'Abril', '05': 'Maio', '06': 'Junho', '07': 'Julho', '08': 'Agosto', '09': 'Setembro', '10': 'Outubro', '11': 'Novembro', '12': 'Dezembro', '03.1': '1º Trimestre', '06.1': '2º Trimestre', '09.1': '3º Trimestre', '12.1': '4º Trimestre'}
		ano_atual = int(date.today().strftime('%Y'))
		anos = [str(x) for x in range(ano_atual - 7, ano_atual + 3)]
		meses = periodos.keys()
		if flag[4:] in meses and flag[:4] in anos:
			self._flag = flag
			self._periodo = periodos[flag[4:]] + ' de ' + flag[:4]
			return True
		return False

	@property
	def documento(self) -> Arquivo:
		return self._documento

	@documento.setter
	def documento(self, documento: str|tuple[str, bool]) -> bool:
		criptografa = False
		try:
			documento, criptografa = documento
		except ValueError:
			pass
		self._documento = Arquivo(classe=self.__class__.__name__, nome=documento, hash=criptografa)
		return True if documento else None

	@property
	def usuario(self) -> Usuario:
		return self._usuario

	@usuario.setter
	def usuario(self, usuario: Usuario) -> bool:
		if isinstance(usuario, Usuario):
			self._usuario = usuario
			return True
		return False

	@property
	def data(self) -> datetime:
		return self._data

	@data.setter
	def data(self, data: str) -> bool:
		try:
			self._data = datetime(int(data[:4]), int(data[5:7]), int(data[8:10]), int(data[11:13]), int(data[14:16]), int(data[17:19]))
			return True
		except TypeError:
			return False

	@property
	def ativo(self) -> bool:
		return self._ativo

	@ativo.setter
	def ativo(self, ativo: bool) -> bool:
		self._ativo = bool(ativo)
		return True

File: modelos/Financa.php
<?php
	include_once("Usuario.php");
	/**
	 * Classe Finança com atributos simples (ID, período, arquivo PDF, usuário, data e status)
	 * @author Luiz Joaquim Aderaldo Amichi <luizamichi@luizamichi.com.br>
	 * @copyright 2020 Luiz Joaquim Aderaldo Amichi
	 */
	class Financa {
		/**
		 * Variáveis privadas
		 * @var int $id Chave primária
		 * @var string $periodo Período (semanal, mensal, bimestral, trimestral, semestral ou anual)
		 * @var string $flag Valor utilizado para ordenação por período
		 * @var string $arquivo Endereço do arquivo em formato PDF
		 * @var Usuario $usuario Usuário (ID, nome, e-mail, login, senha, permissão e status)
		 * @var DateTime $data Data do registro na ordem ano-mês-dia hora:minuto:segundo
		 * @var boolean $status Inativo (FALSE) ou ativo (TRUE)
		 */
		private $id;
		private $periodo;
		private $flag;
		private $arquivo;
		private $usuario;
		private $data;
		private $status;

		/**
		 * Instancia a classe
		 * @return void
		 */
		public function __construct() {
			$this->id = 0;
			$this->periodo = "";
			$this->flag = date("Ym");
			$this->arquivo = "";
			$this->usuario = new Usuario();
			$this->data = DateTime::createFromFormat("Y-m-d H:i:s", date("Y-m-d H:i:s"));
			$this->status = FALSE;
		}

		/**
		 * Destrói a classe
		 * @return void
		 */
		public function __destruct() {
			$this->id = NULL;
			$this->periodo = NULL;
			$this->flag = NULL;
			$this->arquivo = NULL;
			$this->usuario = NULL;
			$this->data = NULL;
			$this->status = NULL;
		}

		/**
		 * Clona a classe
		 * @return void
		 */
		public function __clone() {
			$this->id = clone $this->id;
			$this->periodo = clone $this->periodo;
			$this->flag = clone $this->flag;
			$this->arquivo = clone $this->arquivo;
			$this->usuario = clone $this->usuario;
			$this->data = clone $this->data;
			$this->status = clone $this->status;
		}

		/**
		 * Retorna a classe em formato de string
		 * @return string Atributos da classe separados por quebra de linha
		 */
		public function __toString() {
			return "ID: " . $this->id .
			"\nPERÍODO: " . $this->periodo .
			"\nFLAG: " . $this->flag .
			"\nARQUIVO: " . $this->arquivo .
			"\nUSUÁRIO: " . $this->usuario->getNome() .
			"\nDATA: " . $this->data->format("d/m/Y - H:i:s") .
			"\nSTATUS: " . ($this->status ? "ATIVO" : "INATIVO") . "\n";
		}

		/**
		 * Atribui um novo valor ao ID
		 * @param int $id Número inteiro positivo
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setId(int $id) {
			if($id > 0) {
				$this->id = $id;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor ao período da finança
		 * @param string $periodo Semanal, mensal, (bi, tri ou se)mestral ou anual com no mínimo 6 e no máximo 32 caracteres
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setPeriodo(string $periodo) {
			$periodo = trim($periodo);
			if(strlen($periodo) >= 6 && strlen($periodo) <= 32) {
				$flag = "";
				switch(mb_substr($periodo, 0, 3)) {
					case "Jan":
						$flag = "01";
						break;
					case "Fev":
						$flag = "02";
						break;
					case "Mar":
						$flag = "03";
						break;
					case "Abr":
						$flag = "04";
						break;
					case "Mai":
						$flag = "05";
						break;
					case "Jun":
						$flag = "06";
						break;
					case "Jul":
						$flag = "07";
						break;
					case "Ago":
						$flag = "08";
						break;
					case "Set":
						$flag = "09";
						break;
					case "Out":
						$flag = "10";
						break;
					case "Nov":
						$flag = "11";
						break;
					case "Dez":
						$flag = "12";
						break;
					case "1º ":
						$flag = "03.1";
						break;
					case "2º ":
						$flag = "06.1";
						break;
					case "3º ":
						$flag = "09.1";
						break;
					case "4º ":
						$flag = "12.1";
						break;
					default:
						$periodo = "";
				}
				$this->flag = substr($periodo, -4) . $flag;
				$this->periodo = $periodo;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor ao arquivo
		 * @param string $arquivo Endereço do arquivo PDF de tamanho <= 3MB com no mínimo 5 caracteres
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setArquivo(string $arquivo) {
			if(strcmp(strtolower(substr($arquivo, -3)), "pdf") == 0 && strlen($arquivo) >= 5) {
				$this->arquivo = $arquivo;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo hash ao arquivo
		 * @param string $arquivo Endereço do arquivo PDF de tamanho <= 3MB com no mínimo 5 e no máximo 64 caracteres
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function hashArquivo(string $arquivo) {
			if(strcmp(strtolower(substr($arquivo, -3)), "pdf") == 0 && strlen($arquivo) >= 5) {
				if(strlen($arquivo) > 33) {
					$this->arquivo = mb_strtolower("uploads/financas/" . date("YmdHis") . substr($arquivo, -33));
				}
				else {
					$this->arquivo = mb_strtolower("uploads/financas/" . date("YmdHis") . $arquivo);
				}
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo usuário
		 * @param Usuario $usuario Usuário responsável pela finança
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setUsuario(Usuario $usuario) {
			if($usuario->valida()) {
				$this->usuario = $usuario;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor à data
		 * @param string $data Data no formato "YYYY-mm-dd H:m:s"
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setData(string $data) {
			$datetime = DateTime::createFromFormat("Y-m-d H:i:s", $data);
			if($data && $datetime->format("Y-m-d H:i:s") === $data) {
				$this->data = $datetime;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor ao status da finança
		 * @param boolean $status Booleano referente ao estado
		 * @return void
		 */
		public function setStatus(bool $status) {
			$this->status = $status;
		}

		/**
		 * Retorna o ID da finança
		 * @return int Inteiro positivo único
		 */
		public function getId() {
			return $this->id;
		}

		/**
		 * Retorna o período da finança
		 * @return string Descrição do período
		 */
		public function getPeriodo() {
			return $this->periodo;
		}

		/**
		 * Retorna a flag da finança
		 * @return string Valor referente a posição ordenada
		 */
		public function getFlag() {
			return $this->flag;
		}

		/**
		 * Retorna o endereço onde está armazenado o arquivo PDF a partir da raiz
		 * @return string Endereço no formato "pasta/arquivo.pdf"
		 */
		public function getArquivo() {
			return $this->arquivo;
		}

		/**
		 * Retorna o usuário que criou a finança
		 * @return Usuario Objeto usuário (ID, nome, e-mail, login, senha, permissão e status)
		 */
		public function getUsuario() {
			return $this->usuario;
		}

		/**
		 * Retorna a data e hora da criação ou edição da finança
		 * @return DateTime Data no formato "YYYY-mm-dd H:m:s"
		 */
		public function getData() {
			return $this->data;
		}

		/**
		 * Retorna o status de atividade da finança
		 * @return boolean Inativo ou ativo
		 */
		public function isStatus() {
			return $this->status;
		}

		/**
		 * Valida a finança certificando todos os atributos
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function valida() {
			if($this->id && $this->periodo && $this->flag && $this->arquivo && $this->usuario->valida() && $this->data && $this->status) {
				return TRUE;
			}
			else {
				return FALSE;
			}
		}
	}
?>